DB/DataObject.php

```diff
@@ -2436,16 +2436,23 @@ function _query($string)
         $t= explode(' ',microtime());
         $_DB_DATAOBJECT['QUERYENDTIME'] = $time = $t[0]+$t[1];
 
+      $maxTries = defined('CIVICRM_DEADLOCK_RETRIES') ? CIVICRM_DEADLOCK_RETRIES : 3;
+      for ($tries = 0;$tries < $maxTries;$tries++) {
+        if ($_DB_driver == 'DB') {
+          try {
+            $result = $DB->query($string);
+          }
+          catch (PEAR_Exception $e) {
+            // If we have caught a deadlock - let it go around the loop until our tries limit is hit.
+            // else rethrow the exception.
+            if (!stristr($e->getCause()->getUserInfo(), 'nativecode=1205') || ($tries + 1) === $maxTries) {
+              throw $e;
+            }
+            CRM_Core_Error::debug_log_message('Retrying after deadlock hit on attempt ' . $tries + 1 . ' at query : ' . $string);
```

Comment:
This didn't work properly for me with PHP 7 - `.` and `+` seem to have the same precendence, so the output was always `1 at query: ...`

Reply:
ah - the wrong thing in the log - so it should be like the line above?